stan/math/rev/mat/functor/algebra_system.hpp

```diff
@@ -0,0 +1,107 @@
+#ifndef STAN_MATH_REV_MAT_FUNCTOR_ALGEBRA_SYSTEM_HPP
+#define STAN_MATH_REV_MAT_FUNCTOR_ALGEBRA_SYSTEM_HPP
+
+#include <stan/math/prim/mat/fun/Eigen.hpp>
+#include <stan/math/prim/mat/fun/dogleg.hpp>
+#include <stan/math/rev/mat/functor/jacobian.hpp>
+#include <iostream>
+#include <string>
+#include <vector>
+
+namespace stan {
+  namespace math {
+
+    /**
+     * A functor that allows us to treat either x or y as
+     * the independent variable. This allows us to
+     * choose with respect to which variable the
+     * jacobian gets computed.
+     */
+    template <typename F, typename T0, typename T1>
+    struct system_functor {
+    private:
+      F f_;
+      Eigen::Matrix<T0, Eigen::Dynamic, 1> x_;
+      Eigen::Matrix<T1, Eigen::Dynamic, 1> y_;
+      std::vector<double> dat_;
+      std::vector<int> dat_int_;
+      std::ostream* msgs_;
+      bool x_is_dv_;
+
+    public:
+      system_functor() { }
+
+      system_functor(const F f,
```

Comment:
If the constructor arguments persist, they can be stored as references in the system functor rather than copied.